File: backend-java/src/main/java/org/egualpam/contexts/observable/order/application/domain/OrderCreated.java
package org.egualpam.contexts.observable.order.application.domain;

import org.egualpam.contexts.observable.shared.application.domain.DomainEvent;
import org.egualpam.contexts.observable.shared.application.domain.DomainEventId;

import java.time.Instant;

public class OrderCreated extends DomainEvent {

    private final DomainEventId id;
    private final Instant occurredOn;

    private OrderCreated(DomainEventId id, Instant occurredOn, Order order) {
        super(order);
        this.id = id;
        this.occurredOn = occurredOn;
    }

    public static OrderCreated create(String id, Instant occurredOn, Order order) {
        return new OrderCreated(new DomainEventId(id), occurredOn, order);
    }

    @Override
    public DomainEventId id() {
        return id;
    }

    @Override
    public Instant occurredOn() {
        return occurredOn;
    }
}

File: backend/src/main/kotlin/org/egualpam/contexts/observable/order/application/domain/OrderCreated.kt
package org.egualpam.contexts.observable.order.application.domain

import org.egualpam.contexts.observable.shared.application.domain.DomainEvent
import org.egualpam.contexts.observable.shared.application.domain.DomainEventId
import java.time.Instant

class OrderCreated(
  private val id: DomainEventId,
  private val occurredOn: Instant,
  order: Order
) :
  DomainEvent(order) {

  companion object {
    fun create(
      id: String,
      occurredOn: Instant,
      order: Order
    ) = OrderCreated(
        DomainEventId(id),
        occurredOn,
        order,
    )
  }

  override fun id() = id
  override fun occurredOn() = occurredOn
}
